File: Dynamic_Programming/november/2_Nov_palindrome_partitioning/2_Nov_palindrome_partitioning.py
from sys import maxsize

def minPalPartition(str1):
    n = len(str1)

    c = [0 for i in range (n+1)]
    p = [[False for i in range(n+1)] for j in range(n+1)]


    for i in range(n):
        p[i][i] = True
    
    for l in range(2,n+1):
        for i in range(n-l+1):
            j = i+l-1
            if l == 2:
                p[i][j] = (str1[i] == str1[j])
            else:
                p[i][j] = (str1[i] == str1[j]) and p[i+1][j-1]
    
    for i in range(n):
        if p[0][i] == True:
            c[i] == 0
        else:
            c[i] = maxsize
            for j in range(i):
                if(p[j+1][i] == True and c[j]+1 < c[i]):
                    c[i] = c[j] + 1
    
    return c[n-1]

str1 = "ababbbabbababa"
print("Min cuts needed for Palindrome Partitioning is",minPalPartition(str1))

File: Dynamic_Programming/november/2_Nov_palindrome_partitioning/2_Nov_palindrome_partitioning.cpp
#include<bits/stdc++.h>

using namespace std;

int minPalPartition(char * str){

    int n = strlen(str);

    int c[n];

    bool p[n][n];

    int i,j,k,l;

    for(i = 0; i < n; i++){
        p[i][i] = true;
    }

    for(l = 2; l <= n; l++){
        for(i = 0; i < n-l+1; i++){
            j = i+l-1;

            // If L is 2, then we just need to compare two characters. Else 
            // need to check two corner characters and value of P[i+1][j-1] 
            if(l == 2){
                p[i][j] = (str[i] == str[j]);
            }
            else
            {
                p[i][j] = (str[i] == str[j]) && p[i+1][j-1];
            }
            
        }
    }
    for(i = 0; i < n; i++){
        if(p[0][i] == true){
            c[i] = 0;
        }
        else
        {
            c[i] = INT16_MAX;
            for(j = 0; j < i; j++){
                if(p[j+1][i] == true && i+c[j] <c[i]){
                    c[i] = 1+c[j];
                }
            }
        }
        
    }
    return c[n-1];
}

int main(){
    char str[] = "ababbbabbababa";
    cout << "Min cuts needed for Palindrome Partitioning is :" << minPalPartition(str);

    return 0;
}